panic or segfault under heavy output

Spun out of #124 

Using current `master`, specifically `992a466dc1852ea127f75f904a1397fb7877e0a7`

1. checkout code
2. `cd ./cmd/fyneterm`
3. `fyne get --dst . .` -- see #125
4. `./usr/local/bin/fyneterm`
5. Inside the fyneterm app, run: `docker run debian:stable-slim /bin/sh -c 'apt-get update ; apt-get install -y lz4 ; lz4cat /var/lib/apt/lists/*_Packages.lz4'`
   - This produces about 1.2 million lines of output very fast, far faster than fyneterm can render it
6. Before it finishes, fynterm panics (full panic trace below)
   ```
   fatal error: concurrent map read and map write

   goroutine 72 [running]:
   internal/runtime/maps.fatal({0xd14d13?, 0xc003d7d810?})
   	/usr/lib/go-1.25/src/runtime/panic.go:1046 +0x18
   fyne.io/fyne/v2/internal/async.(*Map[...]).Load(...)
   	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/async/map_migratedfynedo.go:25
   ```

<details>
<summary>Full panic trace</summary>

```
fatal error: concurrent map read and map write

goroutine 72 [running]:
internal/runtime/maps.fatal({0xd14d13?, 0xc003d7d810?})
	/usr/lib/go-1.25/src/runtime/panic.go:1046 +0x18
fyne.io/fyne/v2/internal/async.(*Map[...]).Load(...)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/async/map_migratedfynedo.go:25
fyne.io/fyne/v2/internal/cache.GetCanvasForObject({0xe08960?, 0xc000e68500?})
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/cache/canvases.go:12 +0x45
fyne.io/fyne/v2/internal/driver/common.CanvasForObject(...)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/common/driver.go:10
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).CanvasForObject(0xc003d7d814?, {0xe08960?, 0xc000e68500?})
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/glfw/driver.go:82 +0x25
fyne.io/fyne/v2/canvas.Refresh({0xe08960, 0xc000e68500})
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/canvas/canvas.go:17 +0x71
fyne.io/fyne/v2/canvas.(*Text).Refresh(...)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/canvas/text.go:71
fyne.io/fyne/v2/widget.(*textGridRowRenderer).setCellRune(0x74c3a57c0108?, 0x2f90808?, 0x47d277?, {0xe026a0, 0xc003b89700}, {0x0, 0x0})
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/textgrid.go:710 +0x57e
fyne.io/fyne/v2/widget.(*textGridRowRenderer).refreshCells(0xc000e66330)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/textgrid.go:784 +0x439
fyne.io/fyne/v2/widget.(*textGridRowRenderer).Refresh(0xc000e66330)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/textgrid.go:869 +0xfb
fyne.io/fyne/v2/widget.(*BaseWidget).Refresh(0xc000c86820?)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/widget.go:123 +0x52
fyne.io/fyne/v2/widget.(*textGridContentRenderer).Refresh(0xc000c86820)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/textgrid.go:548 +0x59
fyne.io/fyne/v2/widget.(*BaseWidget).Refresh(0xe08840?)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/widget.go:123 +0x52
fyne.io/fyne/v2/widget.(*textGridRenderer).Refresh(0xc000c84d80)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/textgrid.go:458 +0x23f
fyne.io/fyne/v2/widget.(*BaseWidget).Refresh(0xc00022c4d0?)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/widget.go:123 +0x52
github.com/fyne-io/terminal/internal/widget.(*TermGrid).refreshBlink(0xc0003ba000, 0xa0?)
	/home/mgl/3psrc/fyne-terminal/internal/widget/termgrid.go:58 +0x5f
github.com/fyne-io/terminal/internal/widget.(*TermGrid).Refresh(...)
	/home/mgl/3psrc/fyne-terminal/internal/widget/termgrid.go:42
github.com/fyne-io/terminal.(*Terminal).scrollDown(0xc00031a908)
	/home/mgl/3psrc/fyne-terminal/output.go:305 +0x2c5
github.com/fyne-io/terminal.handleOutputLineFeed(0xc00031a908)
	/home/mgl/3psrc/fyne-terminal/output.go:326 +0x56
github.com/fyne-io/terminal.(*Terminal).handleOutput(0xc00031a908, {0xc000d20000, 0xfff, 0x8000})
	/home/mgl/3psrc/fyne-terminal/output.go:161 +0x389
github.com/fyne-io/terminal.(*Terminal).run(0xc00031a908)
	/home/mgl/3psrc/fyne-terminal/term.go:394 +0x31a
github.com/fyne-io/terminal.(*Terminal).RunLocalShell(0xc00031a908)
	/home/mgl/3psrc/fyne-terminal/term.go:411 +0x48
main.newTerminalWindow.func5()
	/home/mgl/3psrc/fyne-terminal/cmd/fyneterm/main.go:117 +0x2d
created by main.newTerminalWindow in goroutine 1
	/home/mgl/3psrc/fyne-terminal/cmd/fyneterm/main.go:116 +0x6be

goroutine 1 [runnable, locked to thread]:
fyne.io/fyne/v2/internal/async.(*Map[...]).Delete(...)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/async/map_migratedfynedo.go:14
fyne.io/fyne/v2/internal/cache.Clean.destroyExpiredCanvases.func4(...)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/cache/base.go:91
fyne.io/fyne/v2/internal/async.(*Map[...]).Range(...)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/async/map_migratedfynedo.go:55
fyne.io/fyne/v2/internal/cache.destroyExpiredCanvases(...)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/cache/base.go:89
fyne.io/fyne/v2/internal/cache.Clean(0x1)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/cache/base.go:52 +0x69e
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).drawSingleFrame(0xc000d6dde8?)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/glfw/loop.go:93 +0x19b
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).runGL(0xc00004de88?)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/glfw/loop.go:179 +0x1c9
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).Run(0xc000307760)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/glfw/driver.go:162 +0x72
fyne.io/fyne/v2/app.(*fyneApp).Run(0xc000307810)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/app/app.go:77 +0x102
fyne.io/fyne/v2/internal/driver/glfw.(*window).ShowAndRun(0xc0002dc680)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/glfw/window.go:222 +0x64
main.main()
	/home/mgl/3psrc/fyne-terminal/cmd/fyneterm/main.go:65 +0x102

goroutine 20 [select]:
fyne.io/fyne/v2/internal/async.(*UnboundedChan[...]).processing(0xe0bac0)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/async/chan.go:57 +0x525
created by fyne.io/fyne/v2/internal/async.NewUnboundedChan[...] in goroutine 1
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/async/chan.go:20 +0x10a

goroutine 21 [IO wait, 2 minutes]:
internal/poll.runtime_pollWait(0x74c3a565e600, 0x72)
	/usr/lib/go-1.25/src/runtime/netpoll.go:351 +0x85
internal/poll.(*pollDesc).wait(0xc0003148a0?, 0xc00040fe3b?, 0x1)
	/usr/lib/go-1.25/src/internal/poll/fd_poll_runtime.go:84 +0x27
internal/poll.(*pollDesc).waitRead(...)
	/usr/lib/go-1.25/src/internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Read(0xc0003148a0, {0xc00040fe3b, 0x10000, 0x10000})
	/usr/lib/go-1.25/src/internal/poll/fd_unix.go:165 +0x279
os.(*File).read(...)
	/usr/lib/go-1.25/src/os/file_posix.go:29
os.(*File).Read(0xc0002d9eb8, {0xc00040fe3b?, 0x0?, 0x0?})
	/usr/lib/go-1.25/src/os/file.go:144 +0x4f
github.com/fsnotify/fsnotify.(*Watcher).readEvents(0xc0003261c0)
	/home/mgl/.go/pkg/mod/github.com/fsnotify/fsnotify@v1.7.0/backend_inotify.go:483 +0xc5
created by github.com/fsnotify/fsnotify.NewBufferedWatcher in goroutine 1
	/home/mgl/.go/pkg/mod/github.com/fsnotify/fsnotify@v1.7.0/backend_inotify.go:270 +0x1c5

goroutine 22 [chan receive, 2 minutes]:
fyne.io/fyne/v2/app.watchFile.func1()
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/app/settings_desktop.go:42 +0x66
created by fyne.io/fyne/v2/app.watchFile in goroutine 1
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/app/settings_desktop.go:41 +0xe9

goroutine 23 [select, 2 minutes]:
fyne.io/fyne/v2/internal/async.(*UnboundedChan[...]).processing(0xe0ba20)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/async/chan.go:43 +0xdd
created by fyne.io/fyne/v2/internal/async.NewUnboundedChan[...] in goroutine 1
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/async/chan.go:20 +0x10a

goroutine 24 [chan receive, 2 minutes]:
main.setupListener.func1()
	/home/mgl/3psrc/fyne-terminal/cmd/fyneterm/main.go:30 +0x65
created by main.setupListener in goroutine 1
	/home/mgl/3psrc/fyne-terminal/cmd/fyneterm/main.go:28 +0x90

goroutine 73 [chan receive, 2 minutes]:
fyne.io/fyne/v2/internal/app.(*Lifecycle).RunEventQueue(...)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/app/lifecycle.go:110
created by fyne.io/fyne/v2/app.(*fyneApp).Run in goroutine 1
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/app/app.go:71 +0xbe

goroutine 74 [IO wait, 2 minutes]:
internal/poll.runtime_pollWait(0x74c3a565e200, 0x72)
	/usr/lib/go-1.25/src/runtime/netpoll.go:351 +0x85
internal/poll.(*pollDesc).wait(0xc0003140c0?, 0xc000dcfe3b?, 0x1)
	/usr/lib/go-1.25/src/internal/poll/fd_poll_runtime.go:84 +0x27
internal/poll.(*pollDesc).waitRead(...)
	/usr/lib/go-1.25/src/internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Read(0xc0003140c0, {0xc000dcfe3b, 0x10000, 0x10000})
	/usr/lib/go-1.25/src/internal/poll/fd_unix.go:165 +0x279
os.(*File).read(...)
	/usr/lib/go-1.25/src/os/file_posix.go:29
os.(*File).Read(0xc0002d8008, {0xc000dcfe3b?, 0xc9db20?, 0xc000cee100?})
	/usr/lib/go-1.25/src/os/file.go:144 +0x4f
github.com/fsnotify/fsnotify.(*Watcher).readEvents(0xc0003ae840)
	/home/mgl/.go/pkg/mod/github.com/fsnotify/fsnotify@v1.7.0/backend_inotify.go:483 +0xc5
created by github.com/fsnotify/fsnotify.NewBufferedWatcher in goroutine 1
	/home/mgl/.go/pkg/mod/github.com/fsnotify/fsnotify@v1.7.0/backend_inotify.go:270 +0x1c5

goroutine 75 [chan receive, 2 minutes]:
fyne.io/fyne/v2/app.watchFile.func1()
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/app/settings_desktop.go:42 +0x66
created by fyne.io/fyne/v2/app.watchFile in goroutine 1
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/app/settings_desktop.go:41 +0xe9

goroutine 76 [chan receive, 2 minutes]:
github.com/rymdport/portal/settings.OnSignalSettingChanged(0xc000395fc0)
	/home/mgl/.go/pkg/mod/github.com/rymdport/portal@v0.4.1/settings/changed.go:23 +0x67
fyne.io/fyne/v2/app.watchTheme.func1()
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/app/app_xdg.go:123 +0x9f
created by fyne.io/fyne/v2/app.watchTheme in goroutine 1
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/app/app_xdg.go:117 +0x4f

goroutine 77 [chan receive, 2 minutes]:
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).catchTerm(0xc000307760)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/glfw/driver_desktop.go:193 +0x74
created by fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).Run in goroutine 1
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/glfw/driver.go:161 +0x68

goroutine 82 [chan receive, 2 minutes]:
github.com/godbus/dbus/v5.newConn.func1()
	/home/mgl/.go/pkg/mod/github.com/godbus/dbus/v5@v5.1.0/conn.go:303 +0x2c
created by github.com/godbus/dbus/v5.newConn in goroutine 76
	/home/mgl/.go/pkg/mod/github.com/godbus/dbus/v5@v5.1.0/conn.go:302 +0x4da

goroutine 83 [syscall, 2 minutes]:
os/signal.signal_recv()
	/usr/lib/go-1.25/src/runtime/sigqueue.go:152 +0x29
os/signal.loop()
	/usr/lib/go-1.25/src/os/signal/signal_unix.go:23 +0x13
created by os/signal.Notify.func1.1 in goroutine 77
	/usr/lib/go-1.25/src/os/signal/signal.go:152 +0x1f

goroutine 5 [IO wait, 2 minutes]:
internal/poll.runtime_pollWait(0x74c3a565e000, 0x72)
	/usr/lib/go-1.25/src/runtime/netpoll.go:351 +0x85
internal/poll.(*pollDesc).wait(0xc00017f400?, 0xc000cfa800?, 0x0)
	/usr/lib/go-1.25/src/internal/poll/fd_poll_runtime.go:84 +0x27
internal/poll.(*pollDesc).waitRead(...)
	/usr/lib/go-1.25/src/internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).ReadMsg(0xc00017f400, {0xc000cfa800, 0x10, 0x10}, {0xc000d72028, 0x1000, 0x1000}, 0x40000000)
	/usr/lib/go-1.25/src/internal/poll/fd_unix.go:295 +0x385
net.(*netFD).readMsg(0xc00017f400, {0xc000cfa800?, 0x7?, 0xe19178?}, {0xc000d72028?, 0xbd95c0?, 0xc000302a30?}, 0xc000f72858?)
	/usr/lib/go-1.25/src/net/fd_posix.go:91 +0x31
net.(*UnixConn).readMsg(0xc0003c4018, {0xc000cfa800?, 0xbbc520?, 0xc000d41b58?}, {0xc000d72028?, 0xe19178?, 0x1ad68a0?})
	/usr/lib/go-1.25/src/net/unixsock_posix.go:115 +0x3d
net.(*UnixConn).ReadMsgUnix(0xc0003c4018, {0xc000cfa800?, 0xc000d6bda0?, 0x54ab46?}, {0xc000d72028?, 0x3000001ad68a0?, 0x74c35c4d7de0?})
	/usr/lib/go-1.25/src/net/unixsock.go:143 +0x36
github.com/godbus/dbus/v5.(*oobReader).Read(0xc000d72008, {0xc000cfa800?, 0xc000d6be08?, 0x422045?})
	/home/mgl/.go/pkg/mod/github.com/godbus/dbus/v5@v5.1.0/transport_unix.go:21 +0x3c
io.ReadAtLeast({0xdfda60, 0xc000d72008}, {0xc000cfa800, 0x10, 0x10}, 0x10)
	/usr/lib/go-1.25/src/io/io.go:335 +0x8e
io.ReadFull(...)
	/usr/lib/go-1.25/src/io/io.go:354
github.com/godbus/dbus/v5.(*unixTransport).ReadMessage(0xc0002f8000)
	/home/mgl/.go/pkg/mod/github.com/godbus/dbus/v5@v5.1.0/transport_unix.go:91 +0xfa
github.com/godbus/dbus/v5.(*Conn).inWorker(0xc0002383c0)
	/home/mgl/.go/pkg/mod/github.com/godbus/dbus/v5@v5.1.0/conn.go:390 +0x37
created by github.com/godbus/dbus/v5.(*Conn).Auth in goroutine 76
	/home/mgl/.go/pkg/mod/github.com/godbus/dbus/v5@v5.1.0/auth.go:118 +0x88c
```
</details>

It takes several seconds to a minute for it to panic under this load, but it always panics before the command output completes for me.

If I run it with just `go run .`, so that it doesn't have the `migrated_fynedo` build tag, the output is slower, but it still crashes, this time a segfault:
```
~/3psrc/fyne-terminal/cmd/fyneterm$ go run .
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xa1cb1c]

goroutine 1 [running, locked to thread]:
fyne.io/fyne/v2/container.(*featureTheme).Color(0x47d277?, {0xcf9b87?, 0x495c49?}, 0x47d277?)
	<autogenerated>:1 +0x1c
fyne.io/fyne/v2/widget.(*textGridRowRenderer).setCellRune(0x7dc7c4557a78?, 0x47d277?, 0xc0018aa2e0?, {0xe09138, 0xc000336240}, {0x0, 0x0})
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/textgrid.go:680 +0x171
```
<details>
<summary>Full segfault trace</summary>

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xa1cb1c]

goroutine 1 [running, locked to thread]:
fyne.io/fyne/v2/container.(*featureTheme).Color(0x47d277?, {0xcf9b87?, 0x495c49?}, 0x47d277?)
	<autogenerated>:1 +0x1c
fyne.io/fyne/v2/widget.(*textGridRowRenderer).setCellRune(0x7dc7c4557a78?, 0x47d277?, 0xc0018aa2e0?, {0xe09138, 0xc000336240}, {0x0, 0x0})
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/textgrid.go:680 +0x171
fyne.io/fyne/v2/widget.(*textGridRowRenderer).refreshCells(0xc000c57860)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/textgrid.go:795 +0x50d
fyne.io/fyne/v2/widget.(*textGridRowRenderer).Refresh(0xc000c57860)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/textgrid.go:869 +0xfb
fyne.io/fyne/v2/widget.(*BaseWidget).Refresh(0xc000876820?)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/widget.go:123 +0x52
fyne.io/fyne/v2/widget.(*textGridContentRenderer).Refresh(0xc000876820)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/textgrid.go:548 +0x59
fyne.io/fyne/v2/widget.(*BaseWidget).Refresh(0xe0f220?)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/widget.go:123 +0x52
fyne.io/fyne/v2/widget.(*textGridRenderer).Refresh(0xc000874de0)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/textgrid.go:458 +0x23f
fyne.io/fyne/v2/widget.(*BaseWidget).Refresh(0x1?)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/widget.go:123 +0x52
fyne.io/fyne/v2/widget.(*TextGrid).Resize(0xc0002bc180, {0x13007b0?, 0x0?})
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/widget/textgrid.go:149 +0x25
github.com/fyne-io/terminal.(*render).Layout(0xe100c8?, {0x342248?, 0xc0?})
	/home/mgl/3psrc/fyne-terminal/render.go:19 +0x1c
fyne.io/fyne/v2/internal/driver/common.(*Canvas).updateLayout(0xe0f160?, {0xe0f160?, 0xc000342248})
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/common/canvas.go:558 +0x102
fyne.io/fyne/v2/internal/driver/common.(*Canvas).EnsureMinSize.func1(0xc000338f00, {0x1?, 0x0?})
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/common/canvas.go:101 +0xdf
fyne.io/fyne/v2/internal/driver/common.(*Canvas).walkTree.func2({0xe0f0a0?, 0xc00021c1e0?}, {0xe10000?, 0x0?}, {0xe0f160?, 0xc000342248?})
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/common/canvas.go:477 +0xda
fyne.io/fyne/v2/internal/driver.walkObjectTree({0xe0f160, 0xc000342248}, 0x0, {0xe0f040, 0xc000336140}, {0xb398e8?, 0xc0?}, {0x47d0c8?, 0x0?}, {0x6c716d00?, ...}, ...)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/util.go:199 +0x3c2
fyne.io/fyne/v2/internal/driver.walkObjectTree.func1(...)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/util.go:178
fyne.io/fyne/v2/internal/driver.walkObjectTree({0xe0f040, 0xc000336140}, 0x0, {0xe0f040, 0xc000336180}, {0x875620?, 0xc0?}, {0x47db9a?, 0x0?}, {0xe10198?, ...}, ...)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/util.go:192 +0x36c
fyne.io/fyne/v2/internal/driver.walkObjectTree.func1(...)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/util.go:178
fyne.io/fyne/v2/internal/driver.walkObjectTree({0xe0f040, 0xc000336180}, 0x0, {0xe0f1c0, 0xc000252150}, {0x418c0000?, 0x41866667?}, {0xe0f160?, 0x0?}, {0x1?, ...}, ...)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/util.go:192 +0x36c
fyne.io/fyne/v2/internal/driver.walkObjectTree.func1(...)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/util.go:178
fyne.io/fyne/v2/internal/driver.walkObjectTree({0xe0f1c0, 0xc000252150}, 0x0, {0x0, 0x0}, {0x187c470?, 0xc0?}, {0x0?, 0x0?}, {0x418c0000?, ...}, ...)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/util.go:192 +0x36c
fyne.io/fyne/v2/internal/driver.WalkVisibleObjectTree({0xe0f1c0?, 0xc000252150?}, 0xe143e0?, 0xe10198?)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/util.go:134 +0x46
fyne.io/fyne/v2/internal/driver/common.(*Canvas).walkTree(0xa0e5ae?, 0xe10198?, 0xc000252150?, 0xc000b39c00?)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/common/canvas.go:483 +0xa9
fyne.io/fyne/v2/internal/driver/common.(*Canvas).WalkTrees(0xc0001e2c60, 0x0, 0xc000b39c50)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/common/canvas.go:399 +0x36
fyne.io/fyne/v2/internal/driver/common.(*Canvas).EnsureMinSize(0xc0001e2c60)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/common/canvas.go:127 +0xcb
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).repaintWindow(0x0?, 0xc000180000)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/glfw/loop.go:203 +0x27
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).drawSingleFrame.func2()
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/glfw/loop.go:90 +0x30
fyne.io/fyne/v2/internal/driver/glfw.(*window).RunWithContext(0xa75865?, 0xc000b39d48)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/glfw/window.go:902 +0x36
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).drawSingleFrame(0xc000b39de8?)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/glfw/loop.go:89 +0x17b
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).runGL(0xc00004de88?)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/glfw/loop.go:179 +0x1c9
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).Run(0xc00032f8c0)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/glfw/driver.go:162 +0x72
fyne.io/fyne/v2/app.(*fyneApp).Run(0xc00032f970)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/app/app.go:77 +0x102
fyne.io/fyne/v2/internal/driver/glfw.(*window).ShowAndRun(0xc000180000)
	/home/mgl/.go/pkg/mod/fyne.io/fyne/v2@v2.6.1/internal/driver/glfw/window.go:222 +0x64
main.main()
	/home/mgl/3psrc/fyne-terminal/cmd/fyneterm/main.go:65 +0x102
exit status 2
```
</details>

--- a/output.go
+++ b/output.go
@@ -251,9 +251,6 @@ func (t *Terminal) handleOutputChar(r rune) {
 		t.cursorCol = 0
 		handleOutputLineFeed(t)
 	}
-	for len(t.content.Rows)-1 < t.cursorRow {
-		t.content.Rows = append(t.content.Rows, widget.TextGridRow{})
-	}
 
 	var cellStyle widget.TextGridStyle
 	cellStyle = &widget.CustomTextGridStyle{FGColor: t.currentFG, BGColor: t.currentBG}
@@ -264,7 +261,17 @@ func (t *Terminal) handleOutputChar(r rune) {
 	row, col := t.cursorRow, t.cursorCol
 	cell := widget.TextGridCell{Rune: r, Style: cellStyle}
 	fyne.Do(func() {
+		oldLen := 0
+		if len(t.content.Rows) > row {
+			oldLen = len(t.content.Rows[row].Cells)
+		}
 		t.content.SetCell(row, col, cell)
+
+		for i := oldLen; i < col; i++ {
+			if t.content.Rows[row].Cells[i].Rune == 0 {
+				t.content.Rows[row].Cells[i].Rune = ' '
+			}
+		}
 	})
 	t.cursorCol++
 }

--- a/output.go
+++ b/output.go
@@ -257,17 +257,15 @@ func (t *Terminal) handleOutputChar(r rune) {
 
 	var cellStyle widget.TextGridStyle
 	cellStyle = &widget.CustomTextGridStyle{FGColor: t.currentFG, BGColor: t.currentBG}
-	for len(t.content.Rows[t.cursorRow].Cells)-1 < t.cursorCol {
-		newCell := widget.TextGridCell{
-			Rune:  ' ',
-			Style: cellStyle,
-		}
-		t.content.Rows[t.cursorRow].Cells = append(t.content.Rows[t.cursorRow].Cells, newCell)
-	}
 	if t.blinking {
 		cellStyle = widget2.NewTermTextGridStyle(t.currentFG, t.currentBG, highlightBitMask, t.blinking)
 	}
-	t.content.SetCell(t.cursorRow, t.cursorCol, widget.TextGridCell{Rune: r, Style: cellStyle})
+
+	row, col := t.cursorRow, t.cursorCol
+	cell := widget.TextGridCell{Rune: r, Style: cellStyle}
+	fyne.Do(func() {
+		t.content.SetCell(row, col, cell)
+	})
 	t.cursorCol++
 }
 
@@ -323,7 +321,7 @@ func handleOutputCarriageReturn(t *Terminal) {
 
 func handleOutputLineFeed(t *Terminal) {
 	if t.cursorRow == t.scrollBottom {
-		t.scrollDown()
+		fyne.Do(t.scrollDown)
 		if t.newLineMode {
 			t.moveCursor(t.cursorRow, 0)
 		}

--- a/term_unix.go
+++ b/term_unix.go
@@ -44,7 +44,7 @@ func (t *Terminal) startPTY() (io.WriteCloser, io.Reader, io.Closer, error) {
 	go func() {
 		for {
 			time.Sleep(time.Millisecond * 250)
-			if time.Now().Sub(lastKeyTime).Seconds() > 0.5 {
+			if time.Since(lastKeyTime).Seconds() > 0.5 {
 				continue
 			}
 			wd, _ := os.Readlink("/proc/" + strconv.Itoa(c.Process.Pid) + "/cwd")

